### DISCUSSION TITLE
Scheduling request failed: releasing RRC connection

### DISCUSSION THREAD

**Discussion by Zplusless** [srsran/srsRAN_Project#298] [General Help]
I set up a testbed with two USRP-B210s (with external GPSDO 10MHz clock) following the toturial [srsRAN gNB with srsUE — srsRAN Project documentation](https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html).  The  two USRP-B210s are connected with SMA cable. The srsran_performance script is executed on both server.

However, the connection is not stable. The srsUE often fails to work after several seconds' continues working with iperf. The error is as follows:
```bash
edge@EDGE:~/srs-config$ ./run_ue.external.sh
[sudo] password for edge:
Active RF plugins: libsrsran_rf_uhd.so
Inactive RF plugins:
Reading configuration file ue_lab.cn.external.conf...

Built in Release mode using commit fa56836b1 on branch master.

Opening 1 channels in RF device=uhd with args=clock= external
Supported RF device list: UHD file
[INFO] [UHD] linux; GNU C++ version 11.4.0; Boost_107400; UHD_4.5.0.HEAD-0-g471af98f
[INFO] [LOGGING] Fastpath logging disabled at runtime.
Opening USRP channels=1, args: type=b200,master_clock_rate=23.04e6
[INFO] [UHD RF] RF UHD Generic instance constructed
[INFO] [B200] Detected Device: B210
[INFO] [B200] Operating over USB 3.
[INFO] [B200] Initialize CODEC control...
[INFO] [B200] Initialize Radio control...
[INFO] [B200] Performing register loopback test...
[INFO] [B200] Register loopback test passed
[INFO] [B200] Performing register loopback test...
[INFO] [B200] Register loopback test passed
[INFO] [B200] Asking for clock rate 23.040000 MHz...
[INFO] [B200] Actually got clock rate 23.040000 MHz.
Setting manual TX/RX offset to 300 samples
Waiting PHY to initialize ... done!
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=6734
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected
PDU Session Establishment successful. IP: 10.45.0.14
RRC NR reconfiguration successful.

RF status: O=0, U=1, L=0
Scheduling request failed: releasing RRC connection...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=4014
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=4174
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=4334
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=4494
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=4654
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=4814
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=4974
Scheduling request failed: releasing RRC connection...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=6574
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=6734
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=6894
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=7054
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=7214
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=7374
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=7534
```
![image](https://github.com/srsran/srsRAN_Project/assets/10062702/4461d874-5db0-456b-9ffc-81c7185e4020)

## My configs:
[gnb_local_cn_external.yml.txt](https://github.com/srsran/srsRAN_Project/files/13292002/gnb_local_cn_external.yml.txt)
[ue_lab.cn.external.conf.txt](https://github.com/srsran/srsRAN_Project/files/13292003/ue_lab.cn.external.conf.txt)
gnb server: Intel(R) Xeon(R) D-1747NTE CPU @ 2.50GHz, 10 cores 20 threads
ue server: Intel(R) Core(TM) i5-10500 CPU @ 3.10GHz, 6 cores 12 threads

## logs:
https://drive.google.com/drive/folders/1BM-ufJg5DzqDE5Bl6Y1v0iN3Q8k_akf8?usp=sharing

**Comment by brendan-mcauliffe** [comment]
Hi @Zplusless based on your other open discussion point, I am assuming you have solved this issue? If so can you let me know so I can close this point.

**Reply by Zplusless** [reply]
Hi @brendan-mcauliffe， this issue is not solved. Everytime the srsUE starts, it works well. But several seconds later, `Scheduling request failed: releasing RRC connection...` occurs and the bitrate falls down to zero (as demonstrated in the fig above).

**Reply by pgawlowicz** [reply]
Could you try to install older version of UHD, e.g. 3.15 or 4.3 on the UE PC?

**Reply by Zplusless** [reply]
> Could you try to install older version of UHD, e.g. 3.15 or 4.3 on the UE PC?

Thanks, I'll try

**Reply by pgawlowicz** [reply]
did it help?

**Reply by Zplusless** [reply]
sorry for the late reply. I have re-compiled UHD and srsUE with UHD version 4.3.0.0. 
Then I add following lines in the gnb config
```
prach:
    prach_config_index: 1           # Sets PRACH config to match what is expected by srsUE
```

The problem still exists.

**Reply by herlesupreeth** [reply]
Can you please post your gNB and UE configuration file along with the logs?

**Reply by herlesupreeth** [reply]
The gNB configuration file you pasted above is missing the following elements (Notice the missing coreset0_index)

```
  pdcch:
    dedicated:
      dci_format_0_1_and_1_1: false
      ss2_type: common
    common:
      ss0_index: 0
      coreset0_index: 6
  prach:
    prach_config_index: 1
```

**Reply by Zplusless** [reply]
Thanks，I'll add these parameters and try again.

BTW, in another discussion about using 20MHz in srsue, `coreset0_index` is set to 12. I tried coreset0_index=12 in 10MHz UE but ue failes to work. So I delete this parameter.  How can I decide the right value of `coreset0_index`?

**Reply by herlesupreeth** [reply]
Well, you can use any value from 0 to 14 in the below table but whenever you change that value you would need to set the corresponding SSB ARFCN in srsUE configuration file. You can use coreset0_index 12 or above only in case of 20Mhz BW. The reason for setting that value to 6 is because the default value set for SSB ARFCN in srsUE corresponds to that.

![image](https://github.com/srsran/srsRAN_Project/assets/4115839/48ce7d91-d20d-4949-90dd-45f3c3dbc003)

**Reply by Zplusless** [reply]
After adding the above parameters, the throughput reduced to around 15Mbps but the connection still breaks down after around 50s.
the logs and configs are as follows:

https://drive.google.com/drive/folders/1xVxmfElfyf20DyKl52e7dIPRyZpxgzuN?usp=sharing

**Reply by herlesupreeth** [reply]
Can you make changes to gNB config file as follows:

```
  pdcch:
    dedicated:
      dci_format_0_1_and_1_1: false
      ss2_type: common
    common:
      ss0_index: 0
      coreset0_index: 0
  prach:
    prach_config_index: 1
```

then, run the gNB and take note of the `dl_ssb_arfcn` printed in the console. And, then alter the `ssb_nr_arfcn` value in the UE configuration file in the below section

```
[rat.nr]
bands = 3
nof_carriers = 1
dl_nr_arfcn = 368500
ssb_nr_arfcn = 367930
```

Let me know how it goes

**Reply by Zplusless** [reply]
1. I changed gnb config file as above
2. read dl_ssb_arfcn=367930 from the console
3. add dl_nr_arfcn = 368500, ssb_nr_arfcn = 367930 to the ue config file.
4. run

got this:
https://drive.google.com/drive/folders/1xUPZvMKZxKqr2vBUz2hEliuCwfGGy6Tl?usp=sharing

**Reply by herlesupreeth** [reply]
Can you elaborate whether the issue was resolved or not? And another thing, please use the latest code from main branch so that if this is a bug it may have been resolved

I would suggest to test the gain as follows in gNB config file

```
  tx_gain: 80                                                   # Transmit gain of the RF might need to adjusted to the given situation.
  rx_gain: 40                                                   # Receive gain of the RF might need to adjusted to the given situation. 
```

also in UE config file (comment out `rx_gain` and increase `tx_gain`)

```
tx_gain = 80
#rx_gain = 40
```

Also, in your UE console output I see the following, which could be due to incorrect `time_adv_nsamples` value. 

![image](https://github.com/srsran/srsRAN_Project/assets/4115839/4aba0485-5241-4a5d-882e-6e32953fc433)

**Reply by Zplusless** [reply]
Didn't notice that the master branch has been updated. I'll try the latest code. Thanks.

**Reply by pgawlowicz** [reply]
@Zplusless did the new release solve the issue?

**Reply by Zplusless** [reply]
We are now working on zmq mode. I will test the latest version on UDH mode next week.

**Reply by Zplusless** [reply]
Sorry, there are some problems with our usrp-b210. The UHD mode could not be tested temporarily.

**Comment by Zplusless** [comment]
Should I add parameters like:
```
max_nof_prb = 106
nof_prb = 106
```
in config file of UE? These are used in UE 20MHz config, but result in UE attachment failure with 10MHz.

**Reply by herlesupreeth** [reply]
No, those are only for 20Mhz. I will take a look at the logs and get back to you

**Reply by Zplusless** [reply]
Thanks!

**Comment by pgawlowicz** [comment]
@Zplusless can we close here?

**Reply by Zplusless** [reply]
Sorry, I'm not sure wheter the problem is solved. Our USRP has malfunctioned and cannot be used for testing

**Reply by pgawlowicz** [reply]
no problem, let us know once you can test it again.

**Reply by mochammadfikri28** [reply]
@Zplusless Has your problem been solved? I have the same problem as you. Can you help me? thankyou
